### DISCUSSION TITLE
How to expand property only present on derived (sub) class? (casting)

### DISCUSSION THREAD

**Discussion by brease-colin** [odata2ts/odata2ts#320] [Q&A]
The issue that I'm trying to solve is explained here: https://stackoverflow.com/questions/61770954/odata-v4-expand-navigation-properties-of-different-derived-types

With the following URL being the result I'm looking for: https://services.odata.org/V4/(S(xjqbds2oavibr01gt1fny24s))/TripPinServiceRW/People?$expand=Trips($expand=PlanItems($expand=Microsoft.OData.SampleService.Models.TripPin.Flight/Airline))

The part I'm looking for is the 'cast' before the /Airline part.

When trying to use 'expand' or 'expanding', Typescript forces me to use a property of the parent class, but what I'd like is: if an object in the collection is of a derived (sub) class, then I'd like to expand an additional property.

Is there a feature I'm missing or could I supply a manual URL or extend the URL manually to at least get the results I need? A (temporary) workaround would be fine as well. :)

**Comment by texttechne** [comment]
You can use [Custom FilterExpressions](https://odata2ts.github.io/docs/query-builder/filtering#custom-filter-expressions) as an escape hatch in this scenario.

**Comment by brease-colin** [comment]
Thanks for the quick response!

It took a bit of time to do it right without it messing up the rest of the query. It seems like combining it with another (regular) expand went wrong and you actually have to filter something, otherwise my server fails because of the empty filter, so I ended up with this, including a filter that is true always in my case:

builder.filter(new QFilterExpression('Id ne 00000000-0000-0000-0000-000000000000;$expand=Agendas($expand=Agenda)**,OData.Domain.Entities.Appointment/Activities**'));

Would be nice if some QExpandExpression or so is added as an option, so no filter has to be used that does nothing. Or was there an other way?

Either way, I'm also running into another issue here, but that's not related I think, I'll look into it some more. It seems some of the linked entities and their property names have the (correct) case in the Typescript interfaces (like activities and activityId), but the resulting data actually resulted in capitalized versions (Activities and ActivityId).

**Comment by texttechne** [comment]
Converted this into a discussion, since you started with a question. Feels more appropriate here....

> Would be nice if some QExpandExpression or so is added as an option, so no filter has to be used that does nothing.

Agreed, quite hacky this way. Would you be so kind to open an own issue only concerned with this `QExpandExpression`?

> Either way, I'm also running into another issue here, but that's not related I think, I'll look into it some more. It seems some of the linked entities and their property names have the (correct) case in the Typescript interfaces (like activities and activityId), but the resulting data actually resulted in capitalized versions (Activities and ActivityId).

Have you turned on `allowRenaming`? Then again, it should actually behave the other way around: In your EDMX you should have PascalCase and in the TS interfaces you would have camelCase...

**Comment by brease-colin** [comment]
I will, thanks!

I did turn on `allowRenaming`. It seems to be an issue with inheritance. I collect 'event occurrences' with an expanded 'event'. All event properties turn out fine (`id`, `eventType`, etc.). However, I have sub types (like `Appointment`) and any property that is part of that sub type is not renamed in the actual data result, while the TypeScript class is. The same error is true for sub objects of the `Appointment` class.

Example data result:

```
{
  "@odata.type": "#OData.Domain.Entities.Appointment",
  "eventType": "Appointment",
  "subject": "Test",
  "message": "",
  "LocationId": null,
  "agendaOccurrences": [
    {
      "eventId": "b0d0dca2-818b-4ed5-b28b-353dde3c2f15",
      "etc": "09bc2f2e-b439-49bc-a72d-9ab78a08a7e4",
    }
  ],
  "Activities": [
    {
      "AppointmentId": "b0d0dca2-818b-4ed5-b28b-353dde3c2f15",
      "ActivityId": "85e6e0a4-bc54-4c57-bfca-83c0899f28a4",
      "Id": "b3a0d68f-b19e-42bd-9ff7-023aa930adbe"
    }
  ]
}
```

To be clear, the only properties in the above data that are in the sub class Appointment are LocationId and Activities.

I will turn that into an issue as well later.

**Reply by texttechne** [reply]
Yes please. But be so cool to add the EDMX part as well. Only the relevant `EntityTypes`, leave out the rest.
